blog (#1)
RobertCraigie/dotfiles: glide/blog.md

> TLDR; [Glide](glide-browser.app) is a Firefox-based, [modal](#modes), and [malleable](https://glide-browser.app/config)[^1] web browser for the motivated tinkerer.

Reviewer: maybe: 
not sure but maybe something like

```suggestion
TLDR; [Glide](glide-browser.app) is a Firefox fork, built to be [modal](#modes) and [hackable](https://glide-browser.app/config)[^1] for the motivated tinkerer.
```

##### invisible-heading

[Download](https://glide-browser.app/#download) - [Docs](https://glide-browser.app/) - [Cookbook](https://glide-browser.app/cookbook) - [Source](https://github.com/glide-browser/glide)

---

<br>

I believe browsers should be hackable, just like your editor.

```typescript

> glide.keymaps.set("normal", "gC", async () => {

Reviewer: Maybe add a comment like ~/.config/glide/glide.ts

  // extract the owner and repo from a url like 'https://github.com/glide-browser/glide'
  const [owner, repo] = glide.ctx.url.pathname.split("/").slice(1, 3);
  if (!owner || !repo) throw new Error("current URL is not a github repo");

  // * clone the current github repo to ~/github.com/$owner/$repo
  // * start kitty with neovim open at the cloned repo
  const repo_path = glide.path.join(glide.path.home_dir, "github.com", owner, repo);
  await glide.process.execute("gh", [ "repo", "clone", glide.ctx.url, repo_path]);
  await glide.process.execute("kitty", ["-d", repo_path, "nvim"], { cwd: repo_path });
}, { description: "open the GitHub repo in the focused tab in Neovim" });
```

> > Example Glide keymapping to clone the GitHub repo in the current tab and open it in [Kitty](https://sw.kovidgoyal.net/kitty/) + [Neovim](https://neovim.io/).

Reviewer: nit: I think both kitty and neovim shouldn't be capitalized, but 🤷

> This snippet captures exactly the kind of functionality I personally want from a browser; extreme customisability so that I can make my daily workflows a joy.

Reviewer: the wording in this sentence is not great, maybe like

```suggestion
This snippet captures the customizability I expect from my browser so that I can make my daily workflows a joy.
```

i might also expand the connection to other tools (e.g vim)

```
In vim, you could easily make these kinds of customizations tailored to what you need by writing vimscript or installing plugins. In glide, you edit `glide.ts`, which gives a nice typed API. <probably a bit more expansion based on what you want to say/editorialize>
```

Author: ah ty that sentence is much better, struggled to figure out the right phrasing

### What's wrong with the status quo?

Most browsers only allow you to customise shortcuts for certain predefined actions, and do not allow you to define your own entirely custom keymappings without going through a web extension; and even then support for writing your own code is very limited.

Existing web extensions[^2] do offer similar functionality to Glide but are artificially restricted in scope by security constraints[^3] imposed on all web extensions.

For example, extensions are completely disabled on certain URIs, e.g. [addons.mozilla.org](https://addons.mozilla.org), the json file viewer, etc. Which means that relying on extension defined mappings on those pages will not work.

> There are also *many* useful APIs that are not provided to web extensions, such as arbitrary file writes, spawning processes, and setting prefs.

Reviewer: I think an interesting angle here is why is it that web extensions are limited.

My take is that web extension limitations are there mostly to protect the user from the extension writers. In Glide, you are trusted to write the code for yourself.

Reviewer: oh you do write this below, me dumb

Reviewer: I do think it'd still be cool to set that point up earlier by telling you that the browser is protecting you from the extension writers

Author: yeah good point, I kinda said something about that a bit further down but probably makes more sense for that to be in here

### How is Glide different?

Glide holistically solves these usability issues by introducing [modes](#modes), and a TypeScript [config](https://glide-browser.app/config) that lets you do _anything_[^4] you'd like to.

You can define custom [key mappings](https://glide-browser.app/keys), access the [web extensions API](https://glide-browser.app/extensions), spawn arbitrary [processes](https://glide-browser.app/api#glide.process), define [macros](https://glide-browser.app/api#glide.keys) and more.

> Glide can support APIs and functionality that will never be supported in web extensions as the security model is flipped. You - the end user - are responsible for writing the config code, so there is no reason to restrict what you can do.

Reviewer: ```suggestion
Glide can support APIs and functionality that will never be supported in web extensions as the security model is flipped. You---the end user---are responsible for writing the config code, so there is no reason to restrict what you can do.
```

multiple dashes in markdown are converted to em dashes i believe which looks more professional

Author: Unfortunately does not work :(

also might not be worth it as there's a crowd of people that say shit like "they have an em dash in their writing that means it was AI generated!!!!!" which doesn't seem worth it

Reviewer: Tbh i don't think the AI thing is a good reason not to use em dashes

> Here's a small example that adds g + c as a key mapping to switch to the calendar tab:

Reviewer: ```suggestion
Here's a small example that adds <key>g</key>+<key>c</key> as a key mapping to switch to the calendar tab:
```

if you can style this it's kinda nice... not sure how bear blog renders this

Author: just gets stripped :(

Reviewer: oh wait its <kbd>

Author: wait wdym I can't even copy that symbol lol

Author: oh kbd

Author: that does work actually but doesn't appear to be styled any different ootb

Reviewer: oh lmfao i did not expect that to be the way that rendered

```typescript
glide.keymaps.set("normal", "gc", async () => {
  const tab = await glide.tabs.get_first({
    url: "https://calendar.google.com/*",
  });
  assert(tab && tab.id);
  await browser.tabs.update(tab.id, { active: true });
}, {
  description: "[g]o to [c]alendar.google.com",
});
```

For more examples, see the [cookbook](https://glide-browser.app/cookbook) or my own [dotfiles](https://github.com/RobertCraigie/dotfiles/tree/main/glide).

> The cherry on top is that all of this is built on _top_ of Firefox, meaning that, if you already use Firefox, then your existing extensions and workflows will still work in Glide.

Reviewer: Can take this or leave it, I think removing `meaning that` and adding full stop here improves readability but you can keep it for author voice, imo.

```suggestion
The cherry on top is that all of this is built on _top_ of Firefox. If you already use Firefox, then your existing extensions and workflows will still work in Glide.
```

Reviewer: I also think there's a small opportunity here maybe to say why firefox is great, but up to you.

### Modes

Glide borrows the concept of modes from (neo)vim, every key mapping you define will be attached to a specific mode. Glide will then switch between modes automatically but you can always switch modes manually with custom key mappings.

The default mode is `normal`, where most key mappings are defined. When you click or focus on an input element, Glide will automatically switch to `insert` mode, so that key mappings don't interfere with entering text.

> If a website doesn't play well with your key mappings you can switch to `ignore` mode by holding Shift and pressing Escape, in this mode the only default key mapping is Shift + Escape to exit ignore mode.

Reviewer: Same key thing here

```suggestion
If a website doesn't play well with your key mappings you can switch to `ignore` mode by holding <key>Shift</key>+<key>Esc</key>, in this mode the only default key mapping is Shift + Escape to exit ignore mode.
```

### Navigating

Glide supports a `hint` mode that lets you operate web pages entirely using the keyboard.

Press `f` to enter `hint` mode and Glide will overlay text [labels](https://glide-browser.app/hints#label-generation) over every [hintable](https://glide-browser.app/hints#hintable-elements) element, e.g. links and buttons. Typing the [label](https://glide-browser.app/hints#label-generation) for a hint will then focus and click the element.

See it in action in the [demo video](https://glide-browser.app/#demo-video).

### Personal favourite features

- `gI` focuses the largest visible input element on the page and it feels like magic every time I use it.

- `<space><space>` opens a tab fuzzy finder, which is invaluable for finding that _one_ tab you always lose.

- `<c-i>` and `<c-o>` are also invaluable for navigating previously open tabs.

- `:repl` for quickly testing out config changes.

- [hints](https://glide-browser.app/hints) for when I don't want to reach for my mouse.

- A [which-key](https://github.com/folke/which-key.nvim) inspired UI for reminding you of the many key mappings.

<br>

> I've been daily driving Glide for ~6 months now and while I'm biased, I love it. It's still in a very early alpha stage but if you'd like to try it out, you can download it on macOS / Linux [here](https://glide-browser.app/#download). I'd recommend checking out the tutorial with `:tutor` to get some bearings, although it's not fully complete yet.

Reviewer: Idk why but i love this paragraph

Author: hmmm just the vibe is off? delete it or rewrite you think?

Reviewer: no i like it

Reviewer: it sounds like robert

Author: oh wait what I completely misread your comment wtf

Author: I thought you said you didn't like it lmfao
